File: management-core/src/main/java/com/gu/management/timing/TimingMetric.java
/*
 * Copyright 2010 Guardian News and Media
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.gu.management.timing;

import com.gu.management.status.StatusWriter;

import javax.xml.stream.XMLStreamException;
import javax.xml.stream.XMLStreamWriter;
import java.util.concurrent.atomic.AtomicLong;

public class TimingMetric extends StatusWriter {

	private AtomicLong totalTimeInMillis = new AtomicLong();
	private AtomicLong count = new AtomicLong();
	private String managementStatusElementName;

	public void recordTimeSpent(long durationInMillis) {
		totalTimeInMillis.addAndGet(durationInMillis);
		count.incrementAndGet();
	}

	public long getTotalTimeInMillis() {
		return totalTimeInMillis.get();
	}

	public long getCount() {
		return count.get();
	}

	public String getManagementStatusElementName() {
		return managementStatusElementName;
	}

	public void setManagementStatusElementName(String name) {
		this.managementStatusElementName = name;
	}

	@Override public void writeStatus(XMLStreamWriter xmlStreamWriter) throws XMLStreamException {
		xmlStreamWriter.writeStartElement(managementStatusElementName);
		xmlStreamWriter.writeStartElement("count");
		xmlStreamWriter.writeCharacters(count.toString());
		xmlStreamWriter.writeEndElement();
		xmlStreamWriter.writeStartElement("totalTimeInMillis");
		xmlStreamWriter.writeCharacters(totalTimeInMillis.toString());
		xmlStreamWriter.writeEndElement();
		xmlStreamWriter.writeEndElement();
    }
}

File: management-lift/src/main/scala/com/gu/management/TimingMetric.scala
package com.gu.management

import java.util.concurrent.atomic.AtomicLong

object TimingMetric {
  val Empty = new TimingMetric("Empty")
}

class TimingMetric(statusElementName: String) {
  private val _totalTimeInMillis = new AtomicLong()
  private val _count = new AtomicLong()

  def recordTimeSpent(durationInMillis: Long) {
    _totalTimeInMillis.addAndGet(durationInMillis)
    _count.incrementAndGet
  }

  private def genericXml = <prefix>
    <count>{_count.get}</count>
    <totalTimeInMillis>{_totalTimeInMillis.get}</totalTimeInMillis>
  </prefix>

  def toXml = genericXml.copy(label = statusElementName)

  def totalTimeInMillis = _totalTimeInMillis.get
  def count = _count.get

  // to use this class, you can write your own wrappers
  // and call recordTimeSpent, or you may use this one
  // if you want.
  def apply[T](block: => T) {
    val s = new StopWatch
    val result = block
    recordTimeSpent(s.elapsed)
  }
}

